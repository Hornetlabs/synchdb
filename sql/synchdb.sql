CREATE EXTENSION synchdb CASCADE;

\d

SELECT synchdb_add_conninfo('mysqlconn','127.0.0.1', 3306, 'mysqluser', 'mysqlpwd', 'inventory', 'postgres', 'inventory.orders,inventory.customers', 'mysql');
SELECT synchdb_add_conninfo('sqlserverconn','127.0.0.1', 1433, 'sa', 'Password!', 'testDB', 'postgres', '', 'sqlserver');
SELECT synchdb_add_conninfo('oracleconn','127.0.0.1', 1521, 'c##dbzuser', 'dbz', 'mydb', 'postgres', '', 'oracle');
SELECT synchdb_add_conninfo('errorconn','127.0.0.1', 1521, 'c##dbzuser', 'dbz', 'mydb', 'postgres', '', 'nonexist');

SELECT name, isactive, data->'hostname', data->'port', data->'user', data->'srcdb', data->'table', data->'connector' FROM synchdb_conninfo;

SELECT synchdb_add_extra_conninfo('mysqlconn', 'verufy_ca', 'keystore1', 'keystorepass', 'truststore1', 'truststorepass');
SELECT synchdb_add_extra_conninfo('sqlserverconn', 'verufy_ca', 'keystore2', 'keystorepass', 'truststore2', 'truststorepass');
SELECT synchdb_add_extra_conninfo('oracleconn', 'verufy_ca', 'keystore3', 'keystorepass', 'truststore3', 'truststorepass');

SELECT synchdb_del_extra_conninfo('mysqlconn');
SELECT synchdb_del_extra_conninfo('sqlserverconn');
SELECT synchdb_del_extra_conninfo('oracleconn');

SELECT name, isactive, data->'hostname', data->'port', data->'user', data->'srcdb', data->'table', data->'connector' FROM synchdb_conninfo;

SELECT data->'ssl_mode', data->'ssl_keystore', data->'ssl_truststore' FROM synchdb_conninfo;

SELECT synchdb_add_objmap('mysqlconn', 'table', 'ext_db1.ext_table1', 'pg_table1');
SELECT synchdb_add_objmap('mysqlconn', 'column', 'ext_db1.ext_table1.ext_column1', 'pg_column1');
SELECT synchdb_add_objmap('mysqlconn', 'datatype', 'int', 'bigint');
SELECT synchdb_add_objmap('mysqlconn', 'datatype', 'ext_db1.ext_table1.ext_column1', 'text');
SELECT synchdb_add_objmap('mysqlconn', 'transform', 'ext_db1.ext_table1.ext_column1', '''>>>>>'' || ''%d'' || ''<<<<<''');

SELECT synchdb_add_objmap('sqlserverconn', 'table', 'ext_db1.ext_table2', 'pg_table2');
SELECT synchdb_add_objmap('sqlserverconn', 'column', 'ext_db1.ext_table2.ext_column1', 'pg_column2');
SELECT synchdb_add_objmap('sqlserverconn', 'datatype', 'nchar', 'test');
SELECT synchdb_add_objmap('sqlserverconn', 'datatype', 'ext_db1.ext_table2.ext_column1', 'datetime');
SELECT synchdb_add_objmap('sqlserverconn', 'transform', 'ext_db1.ext_table2.ext_column1', '''>>>>>'' || ''%d'' || ''<<<<<''');

SELECT synchdb_add_objmap('oracleconn', 'table', 'ext_db1.ext_table3', 'pg_table3');
SELECT synchdb_add_objmap('oracleconn', 'column', 'ext_db1.ext_table3.ext_column1', 'pg_column3');
SELECT synchdb_add_objmap('oracleconn', 'datatype', 'number', 'bigint');
SELECT synchdb_add_objmap('oracleconn', 'datatype', 'ext_db1.ext_table3.ext_column1', 'varchar');
SELECT synchdb_add_objmap('oracleconn', 'transform', 'ext_db1.ext_table3.ext_column1', '''>>>>>'' || ''%d'' || ''<<<<<''');

SELECT synchdb_add_objmap('oracleconn', 'notexit', 'notexist', 'notexist');

SELECT * FROM synchdb_objmap;

SELECT synchdb_del_objmap('mysqlconn', 'table', 'ext_db1.ext_table1');
SELECT synchdb_del_objmap('mysqlconn', 'column', 'ext_db1.ext_table1.ext_column1');
SELECT synchdb_del_objmap('mysqlconn', 'datatype', 'int');
SELECT synchdb_del_objmap('mysqlconn', 'datatype', 'ext_db1.ext_table1.ext_column1');
SELECT synchdb_del_objmap('mysqlconn', 'transform', 'ext_db1.ext_table1.ext_column1');

SELECT synchdb_del_objmap('sqlserverconn', 'table', 'ext_db1.ext_table2');
SELECT synchdb_del_objmap('sqlserverconn', 'column', 'ext_db1.ext_table2.ext_column1');
SELECT synchdb_del_objmap('sqlserverconn', 'datatype', 'nchar');
SELECT synchdb_del_objmap('sqlserverconn', 'datatype', 'ext_db1.ext_table2.ext_column1');
SELECT synchdb_del_objmap('sqlserverconn', 'transform', 'ext_db1.ext_table2.ext_column1');

SELECT synchdb_del_objmap('oracleconn', 'table', 'ext_db1.ext_table3');
SELECT synchdb_del_objmap('oracleconn', 'column', 'ext_db1.ext_table3.ext_column1');
SELECT synchdb_del_objmap('oracleconn', 'datatype', 'number');
SELECT synchdb_del_objmap('oracleconn', 'datatype', 'ext_db1.ext_table3.ext_column1');
SELECT synchdb_del_objmap('oracleconn', 'transform', 'ext_db1.ext_table3.ext_column1');

SELECT * FROM synchdb_objmap;

SELECT synchdb_del_conninfo('mysqlconn');
SELECT synchdb_del_conninfo('sqlserverconn');
SELECT synchdb_del_conninfo('oracleconn');

SELECT name, isactive, data->'hostname', data->'port', data->'user', data->'srcdb', data->'table', data->'connector' FROM synchdb_conninfo;
SELECT * FROM synchdb_objmap;

DROP EXTENSION synchdb;
