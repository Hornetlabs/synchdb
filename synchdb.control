# synchdb postgresql extension
comment = 'synchdb extension'
default_version = '1.0'
module_pathname = '$libdir/synchdb'
relocatable = true
requires = 'pgcrypto'

