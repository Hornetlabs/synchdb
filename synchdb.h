/*
 * synchdb.h
 *
 * Header file for the SynchDB synchronization system
 *
 * This file defines the core structures and functions used by SynchDB
 * to manage database synchronization across different connector types.
 *
 * Key components:
 * - Connector types and states
 * - Shared state structures for different database connectors
 * - Function prototypes for shared memory operations
 * 
 * Copyright (c) 2024 Hornetlabs Technology, Inc.
 *
 */

#ifndef SYNCHDB_SYNCHDB_H_
#define SYNCHDB_SYNCHDB_H_

#include "storage/lwlock.h"

/* Constants */
#define SYNCHDB_CONNINFO_NAME_SIZE 64
#define SYNCHDB_CONNINFO_HOSTNAME_SIZE 256
#define SYNCHDB_CONNINFO_USERNAME_SIZE 64
#define SYNCHDB_CONNINFO_PASSWORD_SIZE 128
#define SYNCHDB_CONNINFO_TABLELIST_SIZE 256
#define SYNCHDB_CONNINFO_RULEFILENAME_SIZE 64
#define SYNCHDB_CONNINFO_DB_NAME_SIZE 64
#define SYNCHDB_MAX_ACTIVE_CONNECTORS 30

#define DEBEZIUM_SHUTDOWN_TIMEOUT_MSEC 100000

#define SYNCHDB_OFFSET_SIZE 256
#define SYNCHDB_ERRMSG_SIZE 256
#define SYNCHDB_SNAPSHOT_MODE_SIZE 32

#define SYNCHDB_DATATYPE_NAME_SIZE 64
#define SYNCHDB_OBJ_NAME_SIZE 128
#define SYNCHDB_OBJ_TYPE_SIZE 32
#define SYNCHDB_TRANSFORM_EXPRESSION_SIZE 256
#define SYNCHDB_JSON_PATH_SIZE 128
#define SYNCHDB_INVALID_BATCH_ID -1

/*
 * ex: 	pg_synchdb/[connector]_[name]_offsets.dat
 * 		pg_synchdb/mysql_mysqlconn_offsets.dat
 */
#define SYNCHDB_OFFSET_FILE_PATTERN "pg_synchdb/%s_%s_offsets.dat"
#define SYNCHDB_SECRET "930e62fb8c40086c23f543357a023c0c"

#define SYNCHDB_CONNINFO_TABLE "synchdb_conninfo"
/* Enumerations */

/**
 * ConnectorType - Enum representing different types of database connectors
 */
typedef enum _connectorType
{
	TYPE_UNDEF = 0,
	TYPE_MYSQL,
	TYPE_ORACLE,
	TYPE_SQLSERVER,
} ConnectorType;

/**
 * ConnectorState - Enum representing different states of a connector
 */
typedef enum _connectorState
{
	STATE_UNDEF = 0,
	STATE_STOPPED,		/* connector is stopped */
	STATE_INITIALIZING,	/* connector is initializing java dbz runner engine */
	STATE_PAUSED,		/* connector paused until commanded to resume */
	STATE_SYNCING,		/* connector is polling changes from dbz engine */
	STATE_PARSING,		/* got a change event, try to parse it */
	STATE_CONVERTING,	/* parsing done, try to convert it to pg */
	STATE_EXECUTING,	/* conversion done, try to execute it on pg */
	STATE_OFFSET_UPDATE,/* in this state when user requests offset update */
	STATE_RESTARTING,	/* connector is restarting with new snapshot mode */
	STATE_MEMDUMP,		/* connector is dumping jvm heap memory info */
} ConnectorState;

/**
 * ConnectorStage - Enum representing different stages of connector
 */
typedef enum _connectorStage
{
	STAGE_UNDEF = 0,
	STAGE_INITIAL_SNAPSHOT,
	STAGE_CHANGE_DATA_CAPTURE,
} ConnectorStage;

/**
 * ConnectorStatistics - Enum representing different statistics of a connector
 */
typedef enum _connectorStatistics
{
	STATS_UNDEF = 0,
	STATS_DDL,
	STATS_DML,
	STATS_READ,
	STATS_CREATE,
	STATS_UPDATE,
	STATS_DELETE,
	STATS_BAD_CHANGE_EVENT,
	STATS_TOTAL_CHANGE_EVENT,
	STATS_BATCH_COMPLETION,
	STATS_AVERAGE_BATCH_SIZE
} ConnectorStatistics;

/**
 * BatchInfo - Structure containing the metadata of a batch change request
 */
typedef struct _BatchInfo
{
	 int batchId;
	 int batchSize;
} BatchInfo;

/**
 * ConnectionInfo - DBZ Connection info. These are put in shared memory so
 * connector background workers can access when they are spawned.
 */
typedef struct _ConnectionInfo
{
	char name[SYNCHDB_CONNINFO_NAME_SIZE];
    char hostname[SYNCHDB_CONNINFO_HOSTNAME_SIZE];
    unsigned int port;
    char user[SYNCHDB_CONNINFO_USERNAME_SIZE];
    char pwd[SYNCHDB_CONNINFO_PASSWORD_SIZE];
	char srcdb[SYNCHDB_CONNINFO_DB_NAME_SIZE];
	char dstdb[SYNCHDB_CONNINFO_DB_NAME_SIZE];
    char table[SYNCHDB_CONNINFO_TABLELIST_SIZE];
    bool active;
    char rulefile[SYNCHDB_CONNINFO_RULEFILENAME_SIZE];
} ConnectionInfo;

/**
 * ExtraConnectionInfo - Extra DBZ Connection info parameters read from the
 * rule file (if specified). These won't be put in shared memory so they
 * are declared as pointers.
 */
typedef struct _ExtraConnectionInfo
{
	char * ssl_mode;
	char * ssl_keystore;
	char * ssl_keystore_pass;
	char * ssl_truststore;
	char * ssl_truststore_pass;
} ExtraConnectionInfo;

/**
 * SynchdbRequest - Structure representing a request to change connector state
 */
typedef struct _SynchdbRequest
{
	ConnectorState reqstate;
	char reqdata[SYNCHDB_ERRMSG_SIZE];
	ConnectionInfo reqconninfo;
} SynchdbRequest;

/**
 * SynchdbRequest - Structure representing a statistic info per connector.
 * If you add new stats values here, make sure to add the same to ConnectorStatistics
 * enum above
 *
 * todo: to be persisted in future
 */
typedef struct _SynchdbStatistics
{
	unsigned long long stats_ddl;				/* number of DDL operations performed */
	unsigned long long stats_dml;				/* number of DML operations performed */
	unsigned long long stats_read;				/* READ events generated during initial snapshot */
	unsigned long long stats_create;			/* INSERT events generated during CDC */
	unsigned long long stats_update;			/* UPDATE events generated during CDC */
	unsigned long long stats_delete;			/* DELETE events generated during CDC */
	unsigned long long stats_bad_change_event;	/* number of bad change events */
	unsigned long long stats_total_change_event;/* number of total change events */
	unsigned long long stats_batch_completion;	/* number of batches completed */
	unsigned long long stats_average_batch_size;/* calculated average batch size: */

	/* todo: more stats to be added */
} SynchdbStatistics;

/**
 *  Structure holding state information for connectors
 */
typedef struct _ActiveConnectors
{
	pid_t pid;
	ConnectorState state;
	ConnectorStage stage;
	ConnectorType type;
	SynchdbRequest req;
	char errmsg[SYNCHDB_ERRMSG_SIZE];
	char dbzoffset[SYNCHDB_OFFSET_SIZE];
	char snapshotMode[SYNCHDB_SNAPSHOT_MODE_SIZE];
	ConnectionInfo conninfo;
	SynchdbStatistics stats;
} ActiveConnectors;

/**
 * SynchdbSharedState - Shared state information for synchdb background worker
 */
typedef struct _SynchdbSharedState
{
	LWLock		lock;		/* mutual exclusion */
	ActiveConnectors connectors[SYNCHDB_MAX_ACTIVE_CONNECTORS];
} SynchdbSharedState;

/* Function prototypes */
const char * get_shm_connector_name(ConnectorType type);
pid_t get_shm_connector_pid(int connectorId);
void set_shm_connector_pid(int connectorId, pid_t pid);
void set_shm_connector_errmsg(int connectorId, const char * err);
const char * get_shm_connector_errmsg(int connectorId);
void set_shm_connector_state(int connectorId, ConnectorState state);
const char * get_shm_connector_state(int connectorId);
void set_shm_dbz_offset(int connectorId);
const char * get_shm_dbz_offset(int connectorId);
ConnectorState get_shm_connector_state_enum(int connectorId);
const char* connectorTypeToString(ConnectorType type);
void set_shm_connector_stage(int connectorId, ConnectorStage stage);
ConnectorStage get_shm_connector_stage_enum(int connectorId);
void increment_connector_statistics(SynchdbStatistics * myStats, ConnectorStatistics which, int incby);

#endif /* SYNCHDB_SYNCHDB_H_ */
