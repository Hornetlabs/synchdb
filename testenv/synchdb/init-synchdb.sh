#!/bin/bash

PGDATA=${HOME}/synchdb-test
PGLOG=${HOME}/logfile
SOCKDIR=/var/run/postgresql

if [ ! -d $SOCKDIR ]; then
	mkdir -p $SOCKDIR
	chown ubuntu:ubuntu /var/run/postgresql
	chmod 775 /var/run/postgresql
fi

if [ -d "$PGDATA" ]; then
	echo "synchdb is already initialized"
	#pg_ctl -D $PGDATA -l $PGLOG start
	exec postgres -D $PGDATA
	exit 0
fi

initdb -D $PGDATA
#echo "unix_socket_directories = '/home/ubuntu/tmp'" >> $PGDATA/postgresql.conf
echo "listen_addresses ='*'" >> $PGDATA/postgresql.conf
#pg_ctl -D $PGDATA -l $PGLOG start
exec postgres -D $PGDATA
