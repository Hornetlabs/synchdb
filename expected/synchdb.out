CREATE EXTENSION synchdb CASCADE;
\d
              List of relations
 Schema |        Name        | Type  | Owner  
--------+--------------------+-------+--------
 public | synchdb_att_view   | view  | ubuntu
 public | synchdb_attribute  | table | ubuntu
 public | synchdb_conninfo   | table | ubuntu
 public | synchdb_objmap     | table | ubuntu
 public | synchdb_state_view | view  | ubuntu
 public | synchdb_stats_view | view  | ubuntu
(6 rows)

SELECT synchdb_add_conninfo('mysqlconn','127.0.0.1', 3306, 'mysqluser', 'mysqlpwd', 'inventory', 'postgres', 'inventory.orders,inventory.customers', 'mysql');
WARNING:  adjusting destination database from postgres to the current database contrib_regression
 synchdb_add_conninfo 
----------------------
                    0
(1 row)

SELECT synchdb_add_conninfo('sqlserverconn','127.0.0.1', 1433, 'sa', 'Password!', 'testDB', 'postgres', '', 'sqlserver');
WARNING:  adjusting destination database from postgres to the current database contrib_regression
 synchdb_add_conninfo 
----------------------
                    0
(1 row)

SELECT synchdb_add_conninfo('oracleconn','127.0.0.1', 1521, 'c##dbzuser', 'dbz', 'mydb', 'postgres', '', 'oracle');
WARNING:  adjusting destination database from postgres to the current database contrib_regression
 synchdb_add_conninfo 
----------------------
                    0
(1 row)

SELECT synchdb_add_conninfo('errorconn','127.0.0.1', 1521, 'c##dbzuser', 'dbz', 'mydb', 'postgres', '', 'nonexist');
WARNING:  adjusting destination database from postgres to the current database contrib_regression
ERROR:  unsupported connector
SELECT name, isactive, data->'hostname', data->'port', data->'user', data->'srcdb', data->'table', data->'connector' FROM synchdb_conninfo;
     name      | isactive |  ?column?   | ?column? |   ?column?   |  ?column?   |                ?column?                |  ?column?   
---------------+----------+-------------+----------+--------------+-------------+----------------------------------------+-------------
 mysqlconn     | f        | "127.0.0.1" | 3306     | "mysqluser"  | "inventory" | "inventory.orders,inventory.customers" | "mysql"
 sqlserverconn | f        | "127.0.0.1" | 1433     | "sa"         | "testDB"    | "null"                                 | "sqlserver"
 oracleconn    | f        | "127.0.0.1" | 1521     | "c##dbzuser" | "mydb"      | "null"                                 | "oracle"
(3 rows)

SELECT synchdb_add_extra_conninfo('mysqlconn', 'verufy_ca', 'keystore1', 'keystorepass', 'truststore1', 'truststorepass');
 synchdb_add_extra_conninfo 
----------------------------
                          0
(1 row)

SELECT synchdb_add_extra_conninfo('sqlserverconn', 'verufy_ca', 'keystore2', 'keystorepass', 'truststore2', 'truststorepass');
 synchdb_add_extra_conninfo 
----------------------------
                          0
(1 row)

SELECT synchdb_add_extra_conninfo('oracleconn', 'verufy_ca', 'keystore3', 'keystorepass', 'truststore3', 'truststorepass');
 synchdb_add_extra_conninfo 
----------------------------
                          0
(1 row)

SELECT synchdb_del_extra_conninfo('mysqlconn');
 synchdb_del_extra_conninfo 
----------------------------
                          0
(1 row)

SELECT synchdb_del_extra_conninfo('sqlserverconn');
 synchdb_del_extra_conninfo 
----------------------------
                          0
(1 row)

SELECT synchdb_del_extra_conninfo('oracleconn');
 synchdb_del_extra_conninfo 
----------------------------
                          0
(1 row)

SELECT name, isactive, data->'hostname', data->'port', data->'user', data->'srcdb', data->'table', data->'connector' FROM synchdb_conninfo;
     name      | isactive |  ?column?   | ?column? |   ?column?   |  ?column?   |                ?column?                |  ?column?   
---------------+----------+-------------+----------+--------------+-------------+----------------------------------------+-------------
 mysqlconn     | f        | "127.0.0.1" | 3306     | "mysqluser"  | "inventory" | "inventory.orders,inventory.customers" | "mysql"
 sqlserverconn | f        | "127.0.0.1" | 1433     | "sa"         | "testDB"    | "null"                                 | "sqlserver"
 oracleconn    | f        | "127.0.0.1" | 1521     | "c##dbzuser" | "mydb"      | "null"                                 | "oracle"
(3 rows)

SELECT data->'ssl_mode', data->'ssl_keystore', data->'ssl_truststore' FROM synchdb_conninfo;
 ?column? | ?column? | ?column? 
----------+----------+----------
          |          | 
          |          | 
          |          | 
(3 rows)

SELECT synchdb_add_objmap('mysqlconn', 'table', 'ext_db1.ext_table1', 'pg_table1');
 synchdb_add_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_add_objmap('mysqlconn', 'column', 'ext_db1.ext_table1.ext_column1', 'pg_column1');
 synchdb_add_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_add_objmap('mysqlconn', 'datatype', 'int', 'bigint');
 synchdb_add_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_add_objmap('mysqlconn', 'datatype', 'ext_db1.ext_table1.ext_column1', 'text');
 synchdb_add_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_add_objmap('mysqlconn', 'transform', 'ext_db1.ext_table1.ext_column1', '''>>>>>'' || ''%d'' || ''<<<<<''');
 synchdb_add_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_add_objmap('sqlserverconn', 'table', 'ext_db1.ext_table2', 'pg_table2');
 synchdb_add_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_add_objmap('sqlserverconn', 'column', 'ext_db1.ext_table2.ext_column1', 'pg_column2');
 synchdb_add_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_add_objmap('sqlserverconn', 'datatype', 'nchar', 'test');
 synchdb_add_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_add_objmap('sqlserverconn', 'datatype', 'ext_db1.ext_table2.ext_column1', 'datetime');
 synchdb_add_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_add_objmap('sqlserverconn', 'transform', 'ext_db1.ext_table2.ext_column1', '''>>>>>'' || ''%d'' || ''<<<<<''');
 synchdb_add_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_add_objmap('oracleconn', 'table', 'ext_db1.ext_table3', 'pg_table3');
 synchdb_add_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_add_objmap('oracleconn', 'column', 'ext_db1.ext_table3.ext_column1', 'pg_column3');
 synchdb_add_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_add_objmap('oracleconn', 'datatype', 'number', 'bigint');
 synchdb_add_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_add_objmap('oracleconn', 'datatype', 'ext_db1.ext_table3.ext_column1', 'varchar');
 synchdb_add_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_add_objmap('oracleconn', 'transform', 'ext_db1.ext_table3.ext_column1', '''>>>>>'' || ''%d'' || ''<<<<<''');
 synchdb_add_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_add_objmap('oracleconn', 'notexit', 'notexist', 'notexist');
ERROR:  unsupported object type notexit
SELECT * FROM synchdb_objmap;
     name      |  objtype  | enabled |             srcobj             |           dstobj           
---------------+-----------+---------+--------------------------------+----------------------------
 mysqlconn     | table     | t       | ext_db1.ext_table1             | pg_table1
 mysqlconn     | column    | t       | ext_db1.ext_table1.ext_column1 | pg_column1
 mysqlconn     | datatype  | t       | int                            | bigint
 mysqlconn     | datatype  | t       | ext_db1.ext_table1.ext_column1 | text
 mysqlconn     | transform | t       | ext_db1.ext_table1.ext_column1 | '>>>>>' || '%d' || '<<<<<'
 sqlserverconn | table     | t       | ext_db1.ext_table2             | pg_table2
 sqlserverconn | column    | t       | ext_db1.ext_table2.ext_column1 | pg_column2
 sqlserverconn | datatype  | t       | nchar                          | test
 sqlserverconn | datatype  | t       | ext_db1.ext_table2.ext_column1 | datetime
 sqlserverconn | transform | t       | ext_db1.ext_table2.ext_column1 | '>>>>>' || '%d' || '<<<<<'
 oracleconn    | table     | t       | ext_db1.ext_table3             | pg_table3
 oracleconn    | column    | t       | ext_db1.ext_table3.ext_column1 | pg_column3
 oracleconn    | datatype  | t       | number                         | bigint
 oracleconn    | datatype  | t       | ext_db1.ext_table3.ext_column1 | varchar
 oracleconn    | transform | t       | ext_db1.ext_table3.ext_column1 | '>>>>>' || '%d' || '<<<<<'
(15 rows)

SELECT synchdb_del_objmap('mysqlconn', 'table', 'ext_db1.ext_table1');
 synchdb_del_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_del_objmap('mysqlconn', 'column', 'ext_db1.ext_table1.ext_column1');
 synchdb_del_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_del_objmap('mysqlconn', 'datatype', 'int');
 synchdb_del_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_del_objmap('mysqlconn', 'datatype', 'ext_db1.ext_table1.ext_column1');
 synchdb_del_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_del_objmap('mysqlconn', 'transform', 'ext_db1.ext_table1.ext_column1');
 synchdb_del_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_del_objmap('sqlserverconn', 'table', 'ext_db1.ext_table2');
 synchdb_del_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_del_objmap('sqlserverconn', 'column', 'ext_db1.ext_table2.ext_column1');
 synchdb_del_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_del_objmap('sqlserverconn', 'datatype', 'nchar');
 synchdb_del_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_del_objmap('sqlserverconn', 'datatype', 'ext_db1.ext_table2.ext_column1');
 synchdb_del_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_del_objmap('sqlserverconn', 'transform', 'ext_db1.ext_table2.ext_column1');
 synchdb_del_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_del_objmap('oracleconn', 'table', 'ext_db1.ext_table3');
 synchdb_del_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_del_objmap('oracleconn', 'column', 'ext_db1.ext_table3.ext_column1');
 synchdb_del_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_del_objmap('oracleconn', 'datatype', 'number');
 synchdb_del_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_del_objmap('oracleconn', 'datatype', 'ext_db1.ext_table3.ext_column1');
 synchdb_del_objmap 
--------------------
                  0
(1 row)

SELECT synchdb_del_objmap('oracleconn', 'transform', 'ext_db1.ext_table3.ext_column1');
 synchdb_del_objmap 
--------------------
                  0
(1 row)

SELECT * FROM synchdb_objmap;
     name      |  objtype  | enabled |             srcobj             |           dstobj           
---------------+-----------+---------+--------------------------------+----------------------------
 mysqlconn     | table     | f       | ext_db1.ext_table1             | pg_table1
 mysqlconn     | column    | f       | ext_db1.ext_table1.ext_column1 | pg_column1
 mysqlconn     | datatype  | f       | int                            | bigint
 mysqlconn     | datatype  | f       | ext_db1.ext_table1.ext_column1 | text
 mysqlconn     | transform | f       | ext_db1.ext_table1.ext_column1 | '>>>>>' || '%d' || '<<<<<'
 sqlserverconn | table     | f       | ext_db1.ext_table2             | pg_table2
 sqlserverconn | column    | f       | ext_db1.ext_table2.ext_column1 | pg_column2
 sqlserverconn | datatype  | f       | nchar                          | test
 sqlserverconn | datatype  | f       | ext_db1.ext_table2.ext_column1 | datetime
 sqlserverconn | transform | f       | ext_db1.ext_table2.ext_column1 | '>>>>>' || '%d' || '<<<<<'
 oracleconn    | table     | f       | ext_db1.ext_table3             | pg_table3
 oracleconn    | column    | f       | ext_db1.ext_table3.ext_column1 | pg_column3
 oracleconn    | datatype  | f       | number                         | bigint
 oracleconn    | datatype  | f       | ext_db1.ext_table3.ext_column1 | varchar
 oracleconn    | transform | f       | ext_db1.ext_table3.ext_column1 | '>>>>>' || '%d' || '<<<<<'
(15 rows)

SELECT synchdb_del_conninfo('mysqlconn');
 synchdb_del_conninfo 
----------------------
                    0
(1 row)

SELECT synchdb_del_conninfo('sqlserverconn');
 synchdb_del_conninfo 
----------------------
                    0
(1 row)

SELECT synchdb_del_conninfo('oracleconn');
 synchdb_del_conninfo 
----------------------
                    0
(1 row)

SELECT name, isactive, data->'hostname', data->'port', data->'user', data->'srcdb', data->'table', data->'connector' FROM synchdb_conninfo;
 name | isactive | ?column? | ?column? | ?column? | ?column? | ?column? | ?column? 
------+----------+----------+----------+----------+----------+----------+----------
(0 rows)

SELECT * FROM synchdb_objmap;
 name | objtype | enabled | srcobj | dstobj 
------+---------+---------+--------+--------
(0 rows)

DROP EXTENSION synchdb;
